tools/osv-linter/internal/checks/record.go

```diff
@@ -19,3 +40,114 @@ func RecordHasAffected(json *gjson.Result, config *Config) (findings []CheckErro
 	}
 	return findings
 }
+
+func UpstreamCheck(json *gjson.Result, config *Config) (findings []CheckError) {
+	upstream := json.Get("upstream")
+	bug := json.Get("id")
+	if !upstream.Exists() {
+		return
+	}
+
+	hasDup, unique := hasDuplicate(upstream)
+	if hasDup {
+		findings = append(findings, CheckError{Message: "Invalid Upstream: upstream should not contain duplicate entries"})
+	}
+
+	if _, exists := unique[bug.String()]; exists {
+		findings = append(findings, CheckError{Message: "Invalid Upstream: upstream should not contain itself"})
+	}
+
+	aliases := json.Get("aliases")
+	if !aliases.Exists() {
+		return
+	}
+
+	hasAliases := false
+	if aliases.IsArray() {
+		for _, bug := range aliases.Array() {
+			if _, exists := unique[bug.String()]; exists {
+				hasAliases = true
+			}
+		}
+	}
+
+	if hasAliases {
+		findings = append(findings, CheckError{Message: "Invalid Upstream: upstream should not contain aliases"})
+	}
```

Comment:
nit: `hasAliases` is not really a good name - this is more `aliasInUpstream`.
But also, you can skip the `hasAliases` variable and just do
```go
if _, exists := unique[bug.String()]; exists {
	findings = append(findings, CheckError{Message: "Invalid Upstream: upstream should not contain aliases"})
	break
}
```

same with `hasRelated` below